File: Behavioral/Memento/python/memento.py
class Originator:
    _state = None

    class Memento:
        def __init__(self, state):
            self._state = state

        def setState(self, state):
            self._state = state

        def getState(self):
            return self._state

    def __init__(self, state = None):
        self._state = state

    def set(self, state):
        if state != None:
            self._state = state

    def createMemento(self):
        return self.Memento(self._state)

    def restore(self, memento):
        self._state = memento.getState()
        return self._state

class Caretaker:
    pointer = 0
    savedStates = []
    def saveMemento(self, element):
        self.pointer  += 1
        self.savedStates.append(element)

    def getMemento(self, index):
        return self.savedStates[index-1]

    def undo(self):
        if self.pointer > 0:
            self.pointer -= 1
            return self.getMemento(self.pointer)
        else:
            return None

    def redo(self):
        if self.pointer < len(self.savedStates):
            self.pointer += 1
            return self.getMemento(self.pointer)
        else:
            return None



caretaker = Caretaker()
originator = Originator()


#Testing code
originator.set("Message")
caretaker.saveMemento(originator.createMemento())
print(originator.restore(caretaker.getMemento(caretaker.pointer)))
originator.set("Typo")
caretaker.saveMemento(originator.createMemento())
print(originator.restore(caretaker.getMemento(caretaker.pointer)))
originator.set(caretaker.undo())
print(originator.restore(caretaker.getMemento(caretaker.pointer)))
originator.set(caretaker.redo())
print(originator.restore(caretaker.getMemento(caretaker.pointer)))

File: Behavioral/Memento/csharp/Memento.cs
using System;
using System.Collections.Generic;

//Originator. The object which's state changes during runtime.
public class ActiveServer
{

    public void SetState(string name, string ip)
    {
        IP = ip;
        Name = name;
        Console.WriteLine($"Active server set to {ip} - {name}");
    }

    public void GetState()
    {
        Console.WriteLine($"Active server is {IP} - {Name}");
    }


    string IP { get; set; }
    string Name { get; set; }

    public Memento GetSnaphot()
    {
        return new Memento() { IP = this.IP, Name = this.Name };
    }

    public void Restore(Memento memento)
    {
        this.IP = memento.IP;
        this.Name = memento.Name;
        Console.WriteLine($"Active server restored to {IP} - {Name}");
    }
}

//Memento. The memory object.
public class Memento
{
    public string IP { get; set; }
    public string Name { get; set; }
}

//Caretaker. Stores memento objects.
public class ServerList
{
    List<Memento> list = new List<Memento>();

    public void Add(Memento memento)
    {
        list.Add(memento);
    }

    public Memento Get(int id)
    {
        return list[id];
    }
}



class Program
{
    static void Main(string[] args)
    {
        ServerList serverList = new ServerList();
        ActiveServer activeServer = new ActiveServer();

        activeServer.SetState("Sicilia", "66.77.88.999");

        serverList.Add(activeServer.GetSnaphot());

        activeServer.SetState("Ireland", "55.66.77.999");

        serverList.Add(activeServer.GetSnaphot());

        activeServer.Restore(serverList.Get(0));

        activeServer.Restore(serverList.Get(1));

        activeServer.SetState("Oregon", "88.99.77.999");

        serverList.Add(activeServer.GetSnaphot());

        activeServer.Restore(serverList.Get(1));

        activeServer.Restore(serverList.Get(2));

        Console.ReadKey();
    }
}
